### PR TITLE
Improve Ra server resilience when log infrastructure encounters faults

### PR DESCRIPTION
**Pull request by kjnilsson**
Various improvements to data safety when log infrastructure processes encounter faults.

In particular there are many improvements and fixes relating to the server -> wal resend protocol including:
* Bug fix to ra_log_cache that would cause most triggered resends result in a ra process crash.
* Dropping fewer messages using the gen_state postpone feature.
* Ra leaders would previously just exit with `wal_down` - now they enter the same await_condition state although with a shorter timeout after which the begin a leader transfer process
* Improved detection and availability when a command is lost on the way to the wal and no further commands are sent.

Also there is a new feature to configure on a per system basis what kind of server recovery should take place when a ra system starts/restarts. There are 3 options:
1. `undefined` : do not restart any ra server
2. `registered`: restart all locally registered servers for the system
3. `mfa`: call a custom function that performs the restart.

This feature will allow dynamically started ra server to be restarted should the ra system crash and restart.

Also improvements to code coverage and refactoring.

Fixes: https://github.com/rabbitmq/ra/issues/416

### CODE REVIEW

**Comment by pjk25** [comment]
@kjnilsson you can rebase now that #431 is merged

File: src/ra_server.erl [hunk 1]
```diff
@@ -518,51 +521,76 @@ handle_leader({PeerId, #append_entries_reply{success = false,
...
+                           [{PeerId, _} | _] ->
+                               [{next_event, cast, {transfer_leadership, PeerId}}]
+                       end,
 
-    {leader, State, Effects};
+            State = State0#{condition =>
+                            #{predicate_fun => fun wal_down_condition/2,
+                              transition_to => leader,
+                              %% TODO: make duration configurable?
+                              timeout => #{duration => 5000,
```
Comment (michaelklishin) [inline src/ra_server.erl:588]: We have seen enough false positives with 5s timeouts that I'd suggest making it 10s even though any hardcoded default will be wrong/suboptimal for certain installations.

**Review by michaelklishin — APPROVED** [review]
I have pushed some mostly cosmetic changes:

 * Less logging
 * Clarified a few comments

File: src/ra_server.erl (hunk 1 above)
Comment (michaelklishin) [reply src/ra_server.erl:588]: Actually, this should be enough time for `wal_down_condition/2` to observe a WAL process change, and even if it is observed at a later point, the resending can still do its job.

Comment (kjnilsson) [reply src/ra_server.erl:588]: yeah I think 5s is fine, it does affect - if we see ill effects we can make it configurable but I doubt it, worse case it  goes back to leader then have to go back to await_condition again.

**Pull request merged**